### PR TITLE
[Rooster-React]: Emoji plugin

### PR DESCRIPTION
**Pull request by juliaroldi**
Import and refactor the Emoji plugin feature from the old Rooster-react package. 
[https://github.com/microsoft/roosterjs-react/tree/master/packages/roosterjs-react-emoji](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2Froosterjs-react%2Ftree%2Fmaster%2Fpackages%2Froosterjs-react-emoji&data=05%7C01%7Cjuliaroldi%40microsoft.com%7Cf9596b7a9e1946c0532108da4f242581%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637909313918039261%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7)

emojiPlugin

### CODE REVIEW

**Comment by JiuqingSong** [comment]
You can now remove SampleColorPickerPluginDataProvider and all related code from demo site since it is conflict with emoji.

File: demo/scripts/controls/MainPane.tsx [hunk 1]
```diff
@@ -49,6 +54,25 @@ const POPOUT_FEATURES = 'menubar=no,statusbar=no,width=1200,height=800';
 const POPOUT_URL = 'about:blank';
 const POPOUT_TARGET = '_blank';
 
+const emojiPaneProps: EmojiStyle = {
```
Comment (JiuqingSong) [inline demo/scripts/controls/MainPane.tsx:57]: Suggest not expose these settings for now, just provide a simple emoji plugin with default settings

File: tools/buildTools/dts.js [hunk 2]
```diff
@@ -260,11 +260,12 @@ function parseImportFrom(content, currentFileName, queue, baseDir, projDir, exte
         } else {
             const imports = matches[1]
                 .split(',')
-                .map(x => x.replace('{', '').replace('}', '').trim())
+                .map(x => x.replace('{', '').replace('}', '').replace('*', '\\*').trim())
```
Comment (JiuqingSong) [inline tools/buildTools/dts.js:263]: is this change still needed?

File: packages-ui/roosterjs-react/lib/emoji/components/EmojiIcon.tsx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    const { emoji, onClick, isSelected, onMouseOver, emojiIconStyle, onFocus, strings, id } = props;
+    const content = strings[emoji.description];
+
+    return (
+        <button
+            id={id}
+            role="option"
+            className={css(classNames.emoji, emojiIconStyle.className, {
+                [classNames.roosterEmojiSelected]: isSelected,
+                [emojiIconStyle.selectedClassName]: isSelected,
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/components/EmojiIcon.tsx:38]: just use classNames should be good enough for now. If need we can add the support for emojiIconStyle later. Same for other customizable options.

File: packages-ui/roosterjs-react/lib/emoji/components/EmojiIcon.tsx [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            },
+        },
+    },
+
+    roosterEmojiSelected: {
+        '@media screen and (-ms-high-contrast: active)': {
+            backgroundColor: 'Highlight',
+            color: 'HighlightText',
+            '-ms-high-contrast-adjust': 'none',
+        },
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/components/EmojiIcon.tsx:93]: Try use color from theme palette. Example: https://github.com/microsoft/roosterjs/blob/681952d93c0abfe7be03e230b6762d7e12d50d71/packages-ui/roosterjs-react/lib/pasteOptions/component/showPasteOptionPane.tsx#L15 Same for other colors if any.

For high contrast, if need maybe we need to do it for all UI components, but not just for emoji. So I think we can remove this for now.

File: packages-ui/roosterjs-react/lib/emoji/components/EmojiNavBar.tsx [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+                },
+            },
+        },
+    },
+    selected: {
+        borderBottom: '2px solid',
+
+        '@media screen and (-ms-high-contrast: active)': {
+            borderBottomColor: 'Highlight',
+        },
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/components/EmojiNavBar.tsx:123]: same

File: packages-ui/roosterjs-react/lib/emoji/components/EmojiNavBar.tsx [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+                backgroundColor: 'Highlight',
+                color: 'HighlightText',
+                '-ms-high-contrast-adjust': 'none',
+            },
+
+            selected: {
+                '@media screen and (-ms-high-contrast: active)': {
+                    border: '1px solid Highlight',
+                    background: 'none',
+                },
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/components/EmojiNavBar.tsx:114]: same

File: packages-ui/roosterjs-react/lib/emoji/components/EmojiNavBar.tsx [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        },
+
+        '&:hover': {
+            cursor: 'default',
+
+            '@media screen and (-ms-high-contrast: active)': {
+                backgroundColor: 'Highlight',
+                color: 'HighlightText',
+                '-ms-high-contrast-adjust': 'none',
+            },
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/components/EmojiNavBar.tsx:108]: same

File: packages-ui/roosterjs-react/lib/emoji/components/EmojiNavBar.tsx [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+                    const friendlyName = strings[key];
+                    return (
+                        <TooltipHost
+                            hostClassName={classNames.navBarTooltip}
+                            content={friendlyName}
+                            key={key}>
+                            <button
+                                className={css(
+                                    classNames.navBarButton,
+                                    navBarStyle.buttonClassName,
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/components/EmojiNavBar.tsx:50]: same for the navBarStyle

File: packages-ui/roosterjs-react/lib/emoji/components/EmojiPane.tsx [hunk 9]
```diff
@@ -0,0 +1,628 @@
...
+            currentFamily: EmojiFamilyKeys.People,
+            search: ':',
+            searchInBox: '',
+        };
+    }
+
+    public static defaultProps: EmojiPaneProps = {
+        onLayoutChanged: NullFunction,
+        onModeChanged: NullFunction,
+    };
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/components/EmojiPane.tsx:130]: can we remove them?

File: packages-ui/roosterjs-react/lib/emoji/type/EmojiStyle.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    className?: string;
+    selectedClassName?: string;
+}
+
+/**
+ * Emoji status bar style options
+ */
+export interface EmojiStatusBarStyle {
+    className?: string;
+}
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/type/EmojiStyle.ts:48]: Can we remove them, and just hard code the styles for now?

File: packages-ui/roosterjs-react/lib/emoji/type/EmojiStrings.ts [hunk 11]
```diff
@@ -0,0 +1,1226 @@
+/**
+ * Emoji Description
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/type/EmojiStrings.ts:2]: can we only do internal export?

File: packages-ui/roosterjs-react/lib/emoji/type/EmojiStyle.ts (hunk 10 above)
Comment (JiuqingSong) [reply packages-ui/roosterjs-react/lib/emoji/type/EmojiStyle.ts:48]: I mean all of these styles and classes

File: packages-ui/roosterjs-react/lib/emoji/utils/getAriaAndDataProp.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+            result[key] = object[key];
+        }
+        return result;
+    }, {} as T);
+}
+
+/**
+ * @internal
+ */
+export function getDataAndAriaProps<T>(props: any): T {
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/utils/getAriaAndDataProp.ts:17]: try avoid using "any"

File: packages-ui/roosterjs-react/lib/emoji/components/showEmojiCallout.tsx [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    cursorRect: DOMRect,
+    strings: Strings,
+    onSelectFromPane: (emoji: Emoji, wordBeforeCursor: string) => void,
+    refreshCalloutDebounced: () => void,
+    onModeChanged: (newMode: EmojiPaneMode) => void,
+    paneRef: (ref: EmojiPane) => void,
+    onHideCallout: () => void,
+    emojiStyle?: EmojiStyle
+) {
+    uiUtilities.renderComponent(
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/components/showEmojiCallout.tsx:89]: this will return a function and you need to save the function and call it to unmount

File: packages-ui/roosterjs-react/lib/emoji/components/EmojiPane.tsx [hunk 14]
```diff
@@ -0,0 +1,628 @@
...
+        position: 'relative',
+    },
+
+    roosterEmojiPane: {
+        padding: '1px',
+        '@media screen and (-ms-high-contrast: active)': {
+            backgroundColor: 'Highlight',
+            color: 'HighlightText',
+            '-ms-high-contrast-adjust': 'none',
+        },
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/components/EmojiPane.tsx:626]: same

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 15]
```diff
@@ -0,0 +1,436 @@
...
+        this.contentDiv = document.createElement('div');
+        document.body.appendChild(this.contentDiv);
+    }
+
+    public initializeContentEditable(contentEditable: HTMLDivElement): void {
+        this.contentEditable = contentEditable;
+    }
+
+    public startEmoji(startingString: string = ':'): void {
+        if (this.editor) {
```
Comment (BryanValverdeU) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:106]: Replace this with !this.editor

Comment (juliaroldi) [reply packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:106]: OMG, this is a typo. Fixing now

File: packages-ui/roosterjs-react/lib/emoji/components/EmojiIcon.tsx [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+    return mergeStyleSets({
+        emoji: {
+            fontSize: '18px',
+            width: '40px',
+            height: '40px',
+            border: '0',
+            outline: 0,
+            position: 'relative',
+            background: palette.themeLight,
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/components/EmojiIcon.tsx:59]: very nice

File: packages-ui/roosterjs-react/lib/emoji/components/EmojiNavBar.tsx [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+                        <TooltipHost
+                            hostClassName={classNames.navBarTooltip}
+                            content={friendlyName}
+                            key={key}>
+                            <button
+                                className={css(classNames.navBarButton, 'emoji-nav-bar-button', {
+                                    [classNames.selected]: selected,
+                                })}
+                                key={key}
+                                onClick={onFamilyClick.bind(this, key)}
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/components/EmojiNavBar.tsx:52]: functional component doesn't have "this".

File: packages-ui/roosterjs-react/lib/emoji/components/EmojiNavBar.tsx [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+
+/**
+ * @internal
+ * Emoji Nav Bar data
+ */
+export interface EmojiNavBarProps {
+    onClick?: (selected: string) => void;
+    currentSelected?: string;
+    getTabId?: (selected: string) => string;
+    strings: Strings;
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/components/EmojiNavBar.tsx:19]: what is this strings for? And you can just use Record<string,string> as type. I don't want to add too many this kind of new types

Comment (JiuqingSong) [reply packages-ui/roosterjs-react/lib/emoji/components/EmojiNavBar.tsx:19]: maybe it needs a better name

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 19]
```diff
@@ -0,0 +1,426 @@
...
+    willHandleEventExclusively?: (event: PluginEvent) => boolean;
+
+    public getName() {
+        return 'Emoji';
+    }
+
+    public dispose() {
+        this.setIsSuggesting(false);
+        if (this.contentDiv) {
+            ReactDOM.unmountComponentAtNode(this.contentDiv);
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:87]: you can cache the disposer function can call it instead. Do not directly use ReactDOM

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 20]
```diff
@@ -0,0 +1,426 @@
...
+//          2: :r
+
+const EMOJI_BEFORE_COLON_REGEX = /([\u0023-\u0039][\u20e3]|[\ud800-\udbff][\udc00-\udfff]|[\u00a9-\u00ae]|[\u2122-\u3299])*([:;][^:]*)/;
+
+/**
+ * Interface to customize the emoji plugin
+ */
+export interface EmojiPluginOptions {
+    searchPlaceholder?: string;
+    searchInputAriaLabel?: string;
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:45]: are these for localized string? If so, please follow the existing pattern for localized strings in other components/plugins. And I'm ok if we do not expose these options for now and add it later.

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 21]
```diff
@@ -0,0 +1,426 @@
...
+            ...EmojiKeywordStrings,
+            ...EmojiFamilyStrings,
+        };
+    }
+
+    setUIUtilities(uiUtilities: UIUtilities) {
+        this.uiUtilities = uiUtilities;
+    }
+
+    willHandleEventExclusively?: (event: PluginEvent) => boolean;
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:78]: remove this

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 22]
```diff
@@ -0,0 +1,426 @@
...
+ * Interface to customize the emoji plugin
+ */
+export interface EmojiPluginOptions {
+    searchPlaceholder?: string;
+    searchInputAriaLabel?: string;
+}
+
+class EmojiPlugin implements ReactEditorPlugin {
+    private editor: Editor;
+    private contentDiv: HTMLDivElement;
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:50]: this is not used at all now. remove it.

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 23]
```diff
@@ -0,0 +1,426 @@
...
+    }
+
+    public initialize(editor: Editor): void {
+        this.editor = editor;
+        const document = editor.getDocument();
+        this.contentDiv = document.createElement('div');
+        document.body.appendChild(this.contentDiv);
+    }
+
+    public initializeContentEditable(contentEditable: HTMLDivElement): void {
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:101]: no one will call this function, right? you can remove it.

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 24]
```diff
@@ -0,0 +1,426 @@
...
+    searchInputAriaLabel?: string;
+}
+
+class EmojiPlugin implements ReactEditorPlugin {
+    private editor: Editor;
+    private contentDiv: HTMLDivElement;
+    private eventHandledOnKeyDown: boolean;
+    private canUndoEmoji: boolean;
+    private isSuggesting: boolean;
+    private contentEditable: HTMLDivElement;
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:54]: remove it

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 25]
```diff
@@ -0,0 +1,426 @@
...
+        document.body.appendChild(this.contentDiv);
+    }
+
+    public initializeContentEditable(contentEditable: HTMLDivElement): void {
+        this.contentEditable = contentEditable;
+    }
+
+    public onPluginEvent(event: PluginEvent): void {
+        const domEvent = event as PluginDomEvent;
+        const keyboardEvent = domEvent.rawEvent as KeyboardEvent;
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:107]: remove them

Comment (JiuqingSong) [reply packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:107]: After you check the event type, you can use the event as its sub type directly

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 26]
```diff
@@ -0,0 +1,426 @@
...
+        if (event.eventType === PluginEventType.KeyDown) {
+            this.eventHandledOnKeyDown = false;
+            if (this.isSuggesting) {
+                this.onKeyDownSuggestingDomEvent(domEvent);
+            } else if (keyboardEvent.which === KeyCodes.backspace && this.canUndoEmoji) {
+                // If KeyDown is backspace and canUndoEmoji, call editor undo
+                this.editor.undo();
+                this.handleEventOnKeyDown(domEvent);
+                this.canUndoEmoji = false;
+            }
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:117]: We should use auto complete api instead. Either fix it or open a bug and add bug number here to track.

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 27]
```diff
@@ -0,0 +1,426 @@
...
+            !this.isModifierKey(keyboardEvent.key)
+        ) {
+            if (this.isSuggesting) {
+                this.onKeyUpSuggestingDomEvent(domEvent);
+            } else {
+                this.onKeyUpDomEvent(domEvent);
+            }
+        } else if (event.eventType === PluginEventType.MouseUp) {
+            // If MouseUp, the emoji cannot be undone
+            this.canUndoEmoji = false;
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:129]: Same here, add "//TODO" with bug number

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 28]
```diff
@@ -0,0 +1,426 @@
...
+     */
+    private onKeyDownSuggestingDomEvent(event: PluginDomEvent): void {
+        // If key is enter, try insert emoji at selection
+        // If key is space and selection is shortcut, try insert emoji
+        const keyboardEvent = event.rawEvent as KeyboardEvent;
+        const selectedEmoji = this.pane.getSelectedEmoji();
+        const wordBeforeCursor = this.getWordBeforeCursor(event);
+
+        let emoji: Emoji;
+        switch (keyboardEvent.which) {
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:147]: Open a bug to port this plugin to use PickerPlugin, and add bug number with "// TODO" here

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 29]
```diff
@@ -0,0 +1,426 @@
...
+                // Otherwise if the search completed and it is a shortcut, insert the first item
+                emoji = this.timer ? matchShortcut(wordBeforeCursor) : selectedEmoji;
+                break;
+            case KeyCodes.left:
+            case KeyCodes.right:
+                const nextIndex = this.pane.navigate(
+                    keyboardEvent.which === KeyCodes.left ? -1 : 1
+                );
+                if (nextIndex >= 0) {
+                    this.contentEditable.setAttribute(
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:174]: use this.editor.setDomAttribute() instead

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 30]
```diff
@@ -0,0 +1,426 @@
...
+
+        const keyboardEvent = event.rawEvent as KeyboardEvent;
+
+        // If this is a character key or backspace
+        // Clear the timer as we will either queue a new timer or stop suggesting
+        if (
+            (keyboardEvent.key.length === 1 && keyboardEvent.which !== KeyCodes.space) ||
+            keyboardEvent.which === KeyCodes.backspace
+        ) {
+            window.clearTimeout(this.timer);
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:223]: replace all window with this.editor.getDocument().defaultView

File: packages-ui/roosterjs-react/lib/emoji/components/showEmojiCallout.tsx [hunk 31]
```diff
@@ -0,0 +1,190 @@
...
+
+/**
+ * @internal
+ * Emoji callout data
+ */
+interface EmojiICallOutPorps {
+    cursorRect: DOMRect;
+    strings: Strings;
+    onSelectFromPane: (emoji: Emoji, wordBeforeCursor: string) => void;
+    refreshCalloutDebounced: () => void;
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/components/showEmojiCallout.tsx:19]: This is never used, remove it

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 32]
```diff
@@ -0,0 +1,426 @@
...
+    private setIsSuggesting(isSuggesting: boolean): void {
+        if (this.isSuggesting === isSuggesting) {
+            return;
+        }
+
+        this.isSuggesting = isSuggesting;
+        if (this.isSuggesting) {
+            this.getCallout();
+            // we need to delay so NVDA will announce the first selection
+            setTimeout(() => {
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:317]: use this.editor.getDocument().defaultView.setTimeout()

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 33]
```diff
@@ -0,0 +1,426 @@
...
+                const { contentEditable } = this;
+                if (contentEditable) {
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:319]: use this.editor.setDomAttribute

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 34]
```diff
@@ -0,0 +1,426 @@
...
+                    contentEditable.setAttribute(AriaAttributes.AutoComplete, 'list');
+                    contentEditable.setAttribute(AriaAttributes.Owns, this.pane.getListId);
+                    contentEditable.setAttribute(
+                        AriaAttributes.ActiveDescendant,
+                        this.pane.getEmojiElementIdByIndex(0)
+                    );
+                }
+            }, 0);
+        } else {
+            ReactDOM.unmountComponentAtNode(this.contentDiv);
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:329]: use disposer()

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 35]
```diff
@@ -0,0 +1,426 @@
...
+
+            this.removeAutoCompleteAriaAttributes();
+        }
+    }
+
+    private removeAutoCompleteAriaAttributes(): void {
+        const { contentEditable } = this;
+        if (contentEditable) {
+            contentEditable.removeAttribute(AriaAttributes.AutoComplete);
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:338]: same

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 36]
```diff
@@ -0,0 +1,426 @@
...
+        node.innerText = emoji.codePoint;
+        if (
+            wordBeforeCursor &&
+            replaceWithNode(this.editor, wordBeforeCursor, node, false /*exactMatch*/)
+        ) {
+            inserted = true;
+            this.canUndoEmoji = true;
+
+            // Update the editor cursor to be after the inserted node
+            window.requestAnimationFrame(() => {
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:358]: this.editor.getDocument().defaultView

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 37]
```diff
@@ -0,0 +1,426 @@
...
+        if (inserted) {
+            this.triggerChangeEvent();
+        }
+
+        this.tryPatchEmojiFont();
+
+        return inserted;
+    }
+
+    private tryPatchEmojiFont(): boolean {
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:377]: If I remember correctly, this is only for IE. So we can safely remove it now

File: packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts [hunk 38]
```diff
@@ -0,0 +1,426 @@
...
+        const wordBeforeCursor = cursorData ? cursorData.getWordBefore() : null;
+        const matches = EMOJI_BEFORE_COLON_REGEX.exec(wordBeforeCursor);
+        return matches && matches.length > 2 && matches[0] === wordBeforeCursor ? matches[2] : null;
+    }
+
+    private triggerChangeEvent(): void {
+        this.editor.triggerContentChangedEvent(ChangeSource.SetContent);
+    }
+
+    private isModifierKey(key: string): boolean {
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts:410]: we have a util function to do this now. You can use it and remove this function

File: packages-ui/roosterjs-react/lib/emoji/type/AriaAttributes.ts [hunk 39]
```diff
@@ -0,0 +1,12 @@
+/**
+ * @internal
+ */
+export const AriaAttributes = {
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/type/AriaAttributes.ts:4]: Use const enum instead

Comment (JiuqingSong) [reply packages-ui/roosterjs-react/lib/emoji/type/AriaAttributes.ts:4]: And this is only used in createEmojiPlugin.ts, so you can move it into that file and no need to export

File: packages-ui/roosterjs-react/lib/emoji/type/Strings.ts [hunk 40]
```diff
@@ -0,0 +1,4 @@
+/**
+ * @internal
+ */
+export type Strings = { [key: string]: string };
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/type/Strings.ts:4]: remove this type.

File: packages-ui/roosterjs-react/lib/emoji/utils/searchEmojis.ts [hunk 41]
```diff
@@ -0,0 +1,44 @@
...
+        return true;
+    });
+
+    return fullMatch.concat(partialMatch);
+}
+
+/**
+ * @internal
+ */
+export function matchShortcut(search: string): Emoji {
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/utils/searchEmojis.ts:33]: Double check if this is conflict with our existing customReplace plugin

**Review by JiuqingSong — COMMENTED** [review]
Generally much better now.

File: tools/buildTools/dts.js [hunk 42]
```diff
@@ -271,6 +271,7 @@ function parseImportFrom(content, currentFileName, queue, baseDir, projDir, exte
             imports.forEach(x => {
                 newContent = newContent.replace(
                     new RegExp(`(\\W|^)(${x})(\\W|$)`, 'gm'),
+
```
Comment (JiuqingSong) [inline tools/buildTools/dts.js:274]: remove this empty line

File: packages-ui/roosterjs-react/lib/emoji/components/EmojiIcon.tsx [hunk 43]
```diff
@@ -0,0 +1,88 @@
...
+            data-is-focusable={true}
+            aria-label={content}
+            aria-selected={isSelected}
+            {...getDataAndAriaProps(props)}>
+            {emoji.codePoint || '…'}
+        </button>
+    );
+}
+
+const getEmojiIconClassNames = memoizeFunction((theme: Theme) => {
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/components/EmojiIcon.tsx:47]: This function will be called every time when EmojiIcon is rendered, which means it will be called hundreds of times. Suggest move it out of this file into EmojiPake.tsx, and just pass in the class name, so it will only be called once.

Comment (JiuqingSong) [reply packages-ui/roosterjs-react/lib/emoji/components/EmojiIcon.tsx:47]: Same for other files. Let's only have one mergeStyleSet() for emoji, and just pass in class names to other inner components

File: packages-ui/roosterjs-react/lib/emoji/components/EmojiPane.tsx [hunk 44]
```diff
@@ -0,0 +1,575 @@
...
+        mode: EmojiPaneMode.Quick,
+        currentEmojiList: CommonEmojis,
+        currentFamily: EmojiFamilyKeys.People,
+        search: ':',
+        searchInBox: '',
+    });
+
+    const listId = `EmojiPane${props.baseId}`;
+
+    const navigate = (
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/components/EmojiPane.tsx:152]: Use React.useCallback for all these functions

File: packages-ui/roosterjs-react/lib/emoji/components/EmojiPane.tsx [hunk 45]
```diff
@@ -0,0 +1,575 @@
...
+const EmojiPane = React.forwardRef(function EmojiPaneFunc(
+    props: EmojiPaneProps,
+    ref: React.Ref<EmojiPane>
+) {
+    let searchBox: ITextField;
+
+    let emojiBody: HTMLElement;
+    let input: HTMLInputElement;
+
+    const [state, setState] = React.useState<EmojiPaneState>({
```
Comment (JiuqingSong) [inline packages-ui/roosterjs-react/lib/emoji/components/EmojiPane.tsx:141]: Need to split the states into each items. e.g.

const [index, setIndex] = React.useState(0);
const [mode, setMode] = React.useState(EmojiPaneMode.Quick);
...
